src/probe/guided_search_iterator.jl

```diff
@@ -0,0 +1,68 @@
+
+@programiterator GuidedSearchIterator(
+    spec::Vector{<:IOExample},
+    symboltable::SymbolTable,
+)
+Base.@kwdef mutable struct GuidedSearchState
+    level::Int64
+    bank::Vector{Vector{RuleNode}}
+    eval_cache::Set
+    iter::NewProgramsIterator
+    next_iter::Union{Tuple{RuleNode, NewProgramsState}, Nothing}
+end
+
+function Base.iterate(iter::GuidedSearchIterator)
+    iterate(iter, GuidedSearchState(
+        level=-1,
+        bank=[],
+        eval_cache=Set(),
+        iter=NewProgramsIterator(0, [], get_grammar(iter.solver)),
+        next_iter=nothing
+    ))
+end
+
+function Base.iterate(iter::GuidedSearchIterator, state::GuidedSearchState)::Union{Tuple{RuleNode, GuidedSearchState}, Nothing}
+    grammar = get_grammar(iter.solver)
+    start_symbol = get_starting_symbol(iter.solver)
+    # wrap in while true to optimize for tail call
```

Comment:
I do not get this. Why would you do this exactly?

Reply:
It could be that the code that actually returns the outcome of the iteration  never returns a program because the programs are skipped because of observational equivalence.
```julia
if return_type(grammar, prog.ind) == start_symbol
    eval_observation, correct_examples = evaluate_program(prog, grammar, iter.spec, iter.symboltable)

    if eval_observation in state.eval_cache # program already cached
        continue # maybe this skips all programs...
    end
    prog._val = (eval_observation, correct_examples)
    
    push!(state.eval_cache, eval_observation) # add result to cache
    push!(state.bank[state.level+1], prog) # add program to bank
    return (prog, state) # return program
end
```


The `while true` forces the function to keep executing until a program is returned. We could have called the function recursively but using an iterative approach is better I think.